src/transformStateWithClones.js

```diff
@@ -7,7 +7,31 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  let currentState = { ...state };
+  const setOfStates = [];
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'clear':
+        currentState = {};
+        setOfStates.push(currentState);
+        break;
+      case 'addProperties':
+        currentState = { ...currentState, ...action.extraData };
+        setOfStates.push(currentState);
+        break;
+      case 'removeProperties':
+        currentState = { ...currentState };
+
+        for (const key of action.keysToRemove) {
+          delete currentState[key];
```

Comment:
Using `delete` on a property is fine, but keep in mind that it might not be the most performant way to remove properties from an object, especially if there are many properties to remove. However, for the current task, this approach is acceptable.